// @angular/core/schematics/migrations/renderer-to-renderer2/util.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/core/schematics/migrations/renderer-to-renderer2/util", ["require", "exports", "typescript"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.findImportSpecifier = exports.findCoreImport = exports.findRendererReferences = void 0;
    const ts = require("typescript");
    /**
     * Finds typed nodes (e.g. function parameters or class properties) that are referencing the old
     * `Renderer`, as well as calls to the `Renderer` methods.
     */
    function findRendererReferences(sourceFile, typeChecker, rendererImport) {
        const typedNodes = new Set();
        const methodCalls = new Set();
        const forwardRefs = new Set();
        const importSpecifier = findImportSpecifier(rendererImport.elements, 'Renderer');
        const forwardRefImport = findCoreImport(sourceFile, 'forwardRef');
        const forwardRefSpecifier = forwardRefImport ? findImportSpecifier(forwardRefImport.elements, 'forwardRef') : null;
        ts.forEachChild(sourceFile, function visitNode(node) {
            if ((ts.isParameter(node) || ts.isPropertyDeclaration(node)) &&
                isReferenceToImport(typeChecker, node.name, importSpecifier)) {
                typedNodes.add(node);
            }
            else if (ts.isAsExpression(node) && isReferenceToImport(typeChecker, node.type, importSpecifier)) {
                typedNodes.add(node);
            }
            else if (ts.isCallExpression(node)) {
                if (ts.isPropertyAccessExpression(node.expression) &&
                    isReferenceToImport(typeChecker, node.expression.expression, importSpecifier)) {
                    methodCalls.add(node);
                }
                else if (
                // If we're dealing with a forwardRef that's returning a Renderer.
                forwardRefSpecifier && ts.isIdentifier(node.expression) &&
                    isReferenceToImport(typeChecker, node.expression, forwardRefSpecifier) &&
                    node.arguments.length) {
                    const rendererIdentifier = findRendererIdentifierInForwardRef(typeChecker, node, importSpecifier);
                    if (rendererIdentifier) {
                        forwardRefs.add(rendererIdentifier);
                    }
                }
            }
            ts.forEachChild(node, visitNode);
        });
        return { typedNodes, methodCalls, forwardRefs };
    }
    exports.findRendererReferences = findRendererReferences;
    /** Finds the import from @angular/core that has a symbol with a particular name. */
    function findCoreImport(sourceFile, symbolName) {
        // Only look through the top-level imports.
        for (const node of sourceFile.statements) {
            if (!ts.isImportDeclaration(node) || !ts.isStringLiteral(node.moduleSpecifier) ||
                node.moduleSpecifier.text !== '@angular/core') {
                continue;
            }
            const namedBindings = node.importClause && node.importClause.namedBindings;
            if (!namedBindings || !ts.isNamedImports(namedBindings)) {
                continue;
            }
            if (findImportSpecifier(namedBindings.elements, symbolName)) {
                return namedBindings;
            }
        }
        return null;
    }
    exports.findCoreImport = findCoreImport;
    /** Finds an import specifier with a particular name, accounting for aliases. */
    function findImportSpecifier(elements, importName) {
        return elements.find(element => {
            const { name, propertyName } = element;
            return propertyName ? propertyName.text === importName : name.text === importName;
        }) ||
            null;
    }
    exports.findImportSpecifier = findImportSpecifier;
    /** Checks whether a node is referring to an import spcifier. */
    function isReferenceToImport(typeChecker, node, importSpecifier) {
        if (importSpecifier) {
            const nodeSymbol = typeChecker.getTypeAtLocation(node).getSymbol();
            const importSymbol = typeChecker.getTypeAtLocation(importSpecifier).getSymbol();
            return !!(nodeSymbol && importSymbol) &&
                nodeSymbol.valueDeclaration === importSymbol.valueDeclaration;
        }
        return false;
    }
    /** Finds the identifier referring to the `Renderer` inside a `forwardRef` call expression. */
    function findRendererIdentifierInForwardRef(typeChecker, node, rendererImport) {
        const firstArg = node.arguments[0];
        if (ts.isArrowFunction(firstArg)) {
            // Check if the function is `forwardRef(() => Renderer)`.
            if (ts.isIdentifier(firstArg.body) &&
                isReferenceToImport(typeChecker, firstArg.body, rendererImport)) {
                return firstArg.body;
            }
            else if (ts.isBlock(firstArg.body) && ts.isReturnStatement(firstArg.body.statements[0])) {
                // Otherwise check if the expression is `forwardRef(() => { return Renderer })`.
                const returnStatement = firstArg.body.statements[0];
                if (returnStatement.expression && ts.isIdentifier(returnStatement.expression) &&
                    isReferenceToImport(typeChecker, returnStatement.expression, rendererImport)) {
                    return returnStatement.expression;
                }
            }
        }
        return null;
    }
});
