// @angular/compiler-cli/src/ngtsc/typecheck/src/checker.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/typecheck/src/checker", ["require", "exports", "tslib", "@angular/compiler-cli/src/ngtsc/file_system", "@angular/compiler-cli/src/ngtsc/shims", "@angular/compiler-cli/src/ngtsc/typecheck/src/api", "@angular/compiler-cli/src/ngtsc/typecheck/src/context", "@angular/compiler-cli/src/ngtsc/typecheck/src/diagnostics"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.TemplateTypeChecker = void 0;
    var tslib_1 = require("tslib");
    var file_system_1 = require("@angular/compiler-cli/src/ngtsc/file_system");
    var shims_1 = require("@angular/compiler-cli/src/ngtsc/shims");
    var api_1 = require("@angular/compiler-cli/src/ngtsc/typecheck/src/api");
    var context_1 = require("@angular/compiler-cli/src/ngtsc/typecheck/src/context");
    var diagnostics_1 = require("@angular/compiler-cli/src/ngtsc/typecheck/src/diagnostics");
    /**
     * Primary template type-checking engine, which performs type-checking using a
     * `TypeCheckingProgramStrategy` for type-checking program maintenance, and the
     * `ProgramTypeCheckAdapter` for generation of template type-checking code.
     */
    var TemplateTypeChecker = /** @class */ (function () {
        function TemplateTypeChecker(originalProgram, typeCheckingStrategy, typeCheckAdapter, config, refEmitter, reflector, compilerHost, priorBuild) {
            this.originalProgram = originalProgram;
            this.typeCheckingStrategy = typeCheckingStrategy;
            this.typeCheckAdapter = typeCheckAdapter;
            this.config = config;
            this.refEmitter = refEmitter;
            this.reflector = reflector;
            this.compilerHost = compilerHost;
            this.priorBuild = priorBuild;
            this.files = new Map();
        }
        /**
         * Reset the internal type-checking program by generating type-checking code from the user's
         * program.
         */
        TemplateTypeChecker.prototype.refresh = function () {
            var e_1, _a, e_2, _b;
            this.files.clear();
            var ctx = new context_1.TypeCheckContext(this.config, this.compilerHost, this.refEmitter, this.reflector);
            try {
                // Typecheck all the files.
                for (var _c = tslib_1.__values(this.originalProgram.getSourceFiles()), _d = _c.next(); !_d.done; _d = _c.next()) {
                    var sf = _d.value;
                    if (sf.isDeclarationFile || shims_1.isShim(sf)) {
                        continue;
                    }
                    var previousResults = this.priorBuild.priorTypeCheckingResultsFor(sf);
                    if (previousResults === null) {
                        // Previous results were not available, so generate new type-checking code for this file.
                        this.typeCheckAdapter.typeCheck(sf, ctx);
                    }
                    else {
                        // Previous results were available, and can be adopted into the current build.
                        ctx.adoptPriorResults(sf, previousResults);
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
                }
                finally { if (e_1) throw e_1.error; }
            }
            var results = ctx.finalize();
            this.typeCheckingStrategy.updateFiles(results.updates, api_1.UpdateMode.Complete);
            try {
                for (var _e = tslib_1.__values(results.perFileData), _f = _e.next(); !_f.done; _f = _e.next()) {
                    var _g = tslib_1.__read(_f.value, 2), file = _g[0], fileData = _g[1];
                    this.files.set(file, fileData);
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
                }
                finally { if (e_2) throw e_2.error; }
            }
            return results;
        };
        /**
         * Retrieve type-checking diagnostics from the given `ts.SourceFile` using the most recent
         * type-checking program.
         */
        TemplateTypeChecker.prototype.getDiagnosticsForFile = function (sf) {
            var path = file_system_1.absoluteFromSourceFile(sf);
            if (!this.files.has(path)) {
                return [];
            }
            var record = this.files.get(path);
            var typeCheckProgram = this.typeCheckingStrategy.getProgram();
            var typeCheckSf = file_system_1.getSourceFileOrError(typeCheckProgram, record.typeCheckFile);
            var rawDiagnostics = [];
            rawDiagnostics.push.apply(rawDiagnostics, tslib_1.__spread(typeCheckProgram.getSemanticDiagnostics(typeCheckSf)));
            if (record.hasInlines) {
                var inlineSf = file_system_1.getSourceFileOrError(typeCheckProgram, path);
                rawDiagnostics.push.apply(rawDiagnostics, tslib_1.__spread(typeCheckProgram.getSemanticDiagnostics(inlineSf)));
            }
            return rawDiagnostics
                .map(function (diag) {
                if (!diagnostics_1.shouldReportDiagnostic(diag)) {
                    return null;
                }
                return diagnostics_1.translateDiagnostic(diag, record.sourceResolver);
            })
                .filter(function (diag) { return diag !== null; })
                .concat(record.genesisDiagnostics);
        };
        return TemplateTypeChecker;
    }());
    exports.TemplateTypeChecker = TemplateTypeChecker;
});
