// @angular/compiler-cli/src/ngtsc/shims/src/adapter.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/shims/src/adapter", ["require", "exports", "tslib", "typescript", "@angular/compiler-cli/src/ngtsc/file_system", "@angular/compiler-cli/src/ngtsc/util/src/typescript", "@angular/compiler-cli/src/ngtsc/shims/src/expando", "@angular/compiler-cli/src/ngtsc/shims/src/util"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.ShimAdapter = void 0;
    var tslib_1 = require("tslib");
    var ts = require("typescript");
    var file_system_1 = require("@angular/compiler-cli/src/ngtsc/file_system");
    var typescript_1 = require("@angular/compiler-cli/src/ngtsc/util/src/typescript");
    var expando_1 = require("@angular/compiler-cli/src/ngtsc/shims/src/expando");
    var util_1 = require("@angular/compiler-cli/src/ngtsc/shims/src/util");
    /**
     * Generates and tracks shim files for each original `ts.SourceFile`.
     *
     * The `ShimAdapter` provides an API that's designed to be used by a `ts.CompilerHost`
     * implementation and allows it to include synthetic "shim" files in the program that's being
     * created. It works for both freshly created programs as well as with reuse of an older program
     * (which already may contain shim files and thus have a different creation flow).
     */
    var ShimAdapter = /** @class */ (function () {
        function ShimAdapter(delegate, tsRootFiles, topLevelGenerators, perFileGenerators, oldProgram) {
            var e_1, _a, e_2, _b, e_3, _c, e_4, _d, e_5, _e;
            this.delegate = delegate;
            /**
             * A map of shim file names to the `ts.SourceFile` generated for those shims.
             */
            this.shims = new Map();
            /**
             * A map of shim file names to existing shims which were part of a previous iteration of this
             * program.
             *
             * Not all of these shims will be inherited into this program.
             */
            this.priorShims = new Map();
            /**
             * File names which are already known to not be shims.
             *
             * This allows for short-circuit returns without the expense of running regular expressions
             * against the filename repeatedly.
             */
            this.notShims = new Set();
            /**
             * The shim generators supported by this adapter as well as extra precalculated data facilitating
             * their use.
             */
            this.generators = [];
            /**
             * A `Set` of shim `ts.SourceFile`s which should not be emitted.
             */
            this.ignoreForEmit = new Set();
            /**
             * Extension prefixes of all installed per-file shims.
             */
            this.extensionPrefixes = [];
            try {
                // Initialize `this.generators` with a regex that matches each generator's paths.
                for (var perFileGenerators_1 = tslib_1.__values(perFileGenerators), perFileGenerators_1_1 = perFileGenerators_1.next(); !perFileGenerators_1_1.done; perFileGenerators_1_1 = perFileGenerators_1.next()) {
                    var gen = perFileGenerators_1_1.value;
                    // This regex matches paths for shims from this generator. The first (and only) capture group
                    // extracts the filename prefix, which can be used to find the original file that was used to
                    // generate this shim.
                    var pattern = "^(.*)\\." + gen.extensionPrefix + "\\.ts$";
                    var regexp = new RegExp(pattern, 'i');
                    this.generators.push({
                        generator: gen,
                        test: regexp,
                        suffix: "." + gen.extensionPrefix + ".ts",
                    });
                    this.extensionPrefixes.push(gen.extensionPrefix);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (perFileGenerators_1_1 && !perFileGenerators_1_1.done && (_a = perFileGenerators_1.return)) _a.call(perFileGenerators_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
            // Process top-level generators and pre-generate their shims. Accumulate the list of filenames
            // as extra input files.
            var extraInputFiles = [];
            try {
                for (var topLevelGenerators_1 = tslib_1.__values(topLevelGenerators), topLevelGenerators_1_1 = topLevelGenerators_1.next(); !topLevelGenerators_1_1.done; topLevelGenerators_1_1 = topLevelGenerators_1.next()) {
                    var gen = topLevelGenerators_1_1.value;
                    var sf = gen.makeTopLevelShim();
                    expando_1.sfExtensionData(sf).isTopLevelShim = true;
                    if (!gen.shouldEmit) {
                        this.ignoreForEmit.add(sf);
                    }
                    var fileName = file_system_1.absoluteFromSourceFile(sf);
                    this.shims.set(fileName, sf);
                    extraInputFiles.push(fileName);
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (topLevelGenerators_1_1 && !topLevelGenerators_1_1.done && (_b = topLevelGenerators_1.return)) _b.call(topLevelGenerators_1);
                }
                finally { if (e_2) throw e_2.error; }
            }
            try {
                // Add to that list the per-file shims associated with each root file. This is needed because
                // reference tagging alone may not work in TS compilations that have `noResolve` set. Such
                // compilations rely on the list of input files completely describing the program.
                for (var tsRootFiles_1 = tslib_1.__values(tsRootFiles), tsRootFiles_1_1 = tsRootFiles_1.next(); !tsRootFiles_1_1.done; tsRootFiles_1_1 = tsRootFiles_1.next()) {
                    var rootFile = tsRootFiles_1_1.value;
                    try {
                        for (var _f = (e_4 = void 0, tslib_1.__values(this.generators)), _g = _f.next(); !_g.done; _g = _f.next()) {
                            var gen = _g.value;
                            extraInputFiles.push(util_1.makeShimFileName(rootFile, gen.suffix));
                        }
                    }
                    catch (e_4_1) { e_4 = { error: e_4_1 }; }
                    finally {
                        try {
                            if (_g && !_g.done && (_d = _f.return)) _d.call(_f);
                        }
                        finally { if (e_4) throw e_4.error; }
                    }
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (tsRootFiles_1_1 && !tsRootFiles_1_1.done && (_c = tsRootFiles_1.return)) _c.call(tsRootFiles_1);
                }
                finally { if (e_3) throw e_3.error; }
            }
            this.extraInputFiles = extraInputFiles;
            // If an old program is present, extract all per-file shims into a map, which will be used to
            // generate new versions of those shims.
            if (oldProgram !== null) {
                try {
                    for (var _h = tslib_1.__values(oldProgram.getSourceFiles()), _j = _h.next(); !_j.done; _j = _h.next()) {
                        var oldSf = _j.value;
                        if (oldSf.isDeclarationFile || !expando_1.isFileShimSourceFile(oldSf)) {
                            continue;
                        }
                        this.priorShims.set(file_system_1.absoluteFromSourceFile(oldSf), oldSf);
                    }
                }
                catch (e_5_1) { e_5 = { error: e_5_1 }; }
                finally {
                    try {
                        if (_j && !_j.done && (_e = _h.return)) _e.call(_h);
                    }
                    finally { if (e_5) throw e_5.error; }
                }
            }
        }
        /**
         * Produce a shim `ts.SourceFile` if `fileName` refers to a shim file which should exist in the
         * program.
         *
         * If `fileName` does not refer to a potential shim file, `null` is returned. If a corresponding
         * base file could not be determined, `undefined` is returned instead.
         */
        ShimAdapter.prototype.maybeGenerate = function (fileName) {
            var e_6, _a;
            // Fast path: either this filename has been proven not to be a shim before, or it is a known
            // shim and no generation is required.
            if (this.notShims.has(fileName)) {
                return null;
            }
            else if (this.shims.has(fileName)) {
                return this.shims.get(fileName);
            }
            // .d.ts files can't be shims.
            if (typescript_1.isDtsPath(fileName)) {
                this.notShims.add(fileName);
                return null;
            }
            try {
                // This is the first time seeing this path. Try to match it against a shim generator.
                for (var _b = tslib_1.__values(this.generators), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var record = _c.value;
                    var match = record.test.exec(fileName);
                    if (match === null) {
                        continue;
                    }
                    // The path matched. Extract the filename prefix without the extension.
                    var prefix = match[1];
                    // This _might_ be a shim, if an underlying base file exists. The base file might be .ts or
                    // .tsx.
                    var baseFileName = file_system_1.absoluteFrom(prefix + '.ts');
                    if (!this.delegate.fileExists(baseFileName)) {
                        // No .ts file by that name - try .tsx.
                        baseFileName = file_system_1.absoluteFrom(prefix + '.tsx');
                        if (!this.delegate.fileExists(baseFileName)) {
                            // This isn't a shim after all since there is no original file which would have triggered
                            // its generation, even though the path is right. There are a few reasons why this could
                            // occur:
                            //
                            // * when resolving an import to an .ngfactory.d.ts file, the module resolution algorithm
                            //   will first look for an .ngfactory.ts file in its place, which will be requested here.
                            // * when the user writes a bad import.
                            // * when a file is present in one compilation and removed in the next incremental step.
                            //
                            // Note that this does not add the filename to `notShims`, so this path is not cached.
                            // That's okay as these cases above are edge cases and do not occur regularly in normal
                            // operations.
                            return undefined;
                        }
                    }
                    // Retrieve the original file for which the shim will be generated.
                    var inputFile = this.delegate.getSourceFile(baseFileName, ts.ScriptTarget.Latest);
                    if (inputFile === undefined || expando_1.isShim(inputFile)) {
                        // Something strange happened here. This case is also not cached in `notShims`, but this
                        // path is not expected to occur in reality so this shouldn't be a problem.
                        return undefined;
                    }
                    // Actually generate and cache the shim.
                    return this.generateSpecific(fileName, record.generator, inputFile);
                }
            }
            catch (e_6_1) { e_6 = { error: e_6_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_6) throw e_6.error; }
            }
            // No generator matched.
            this.notShims.add(fileName);
            return null;
        };
        ShimAdapter.prototype.generateSpecific = function (fileName, generator, inputFile) {
            var priorShimSf = null;
            if (this.priorShims.has(fileName)) {
                // In the previous program a shim with this name already existed. It's passed to the shim
                // generator which may reuse it instead of generating a fresh shim.
                priorShimSf = this.priorShims.get(fileName);
                this.priorShims.delete(fileName);
            }
            var shimSf = generator.generateShimForFile(inputFile, fileName, priorShimSf);
            // Mark the new generated source file as a shim that originated from this generator.
            expando_1.sfExtensionData(shimSf).fileShim = {
                extension: generator.extensionPrefix,
                generatedFrom: file_system_1.absoluteFromSourceFile(inputFile),
            };
            if (!generator.shouldEmit) {
                this.ignoreForEmit.add(shimSf);
            }
            this.shims.set(fileName, shimSf);
            return shimSf;
        };
        return ShimAdapter;
    }());
    exports.ShimAdapter = ShimAdapter;
});
