// @angular/compiler-cli/src/ngtsc/shims/src/expando.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/shims/src/expando", ["require", "exports", "tslib"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.retagTsFile = exports.untagTsFile = exports.retagAllTsFiles = exports.untagAllTsFiles = exports.copyFileShimData = exports.isShim = exports.isFileShimSourceFile = exports.sfExtensionData = exports.isExtended = exports.NgExtension = void 0;
    var tslib_1 = require("tslib");
    /**
     * A `Symbol` which is used to patch extension data onto `ts.SourceFile`s.
     */
    exports.NgExtension = Symbol('NgExtension');
    /**
     * Narrows a `ts.SourceFile` if it has an `NgExtension` property.
     */
    function isExtended(sf) {
        return sf[exports.NgExtension] !== undefined;
    }
    exports.isExtended = isExtended;
    /**
     * Returns the `NgExtensionData` for a given `ts.SourceFile`, adding it if none exists.
     */
    function sfExtensionData(sf) {
        var extSf = sf;
        if (extSf[exports.NgExtension] !== undefined) {
            // The file already has extension data, so return it directly.
            return extSf[exports.NgExtension];
        }
        // The file has no existing extension data, so add it and return it.
        var extension = {
            isTopLevelShim: false,
            fileShim: null,
            originalReferencedFiles: null,
            taggedReferenceFiles: null,
        };
        extSf[exports.NgExtension] = extension;
        return extension;
    }
    exports.sfExtensionData = sfExtensionData;
    /**
     * Check whether `sf` is a per-file shim `ts.SourceFile`.
     */
    function isFileShimSourceFile(sf) {
        return isExtended(sf) && sf[exports.NgExtension].fileShim !== null;
    }
    exports.isFileShimSourceFile = isFileShimSourceFile;
    /**
     * Check whether `sf` is a shim `ts.SourceFile` (either a per-file shim or a top-level shim).
     */
    function isShim(sf) {
        return isExtended(sf) && (sf[exports.NgExtension].fileShim !== null || sf[exports.NgExtension].isTopLevelShim);
    }
    exports.isShim = isShim;
    /**
     * Copy any shim data from one `ts.SourceFile` to another.
     */
    function copyFileShimData(from, to) {
        if (!isFileShimSourceFile(from)) {
            return;
        }
        sfExtensionData(to).fileShim = sfExtensionData(from).fileShim;
    }
    exports.copyFileShimData = copyFileShimData;
    /**
     * For those `ts.SourceFile`s in the `program` which have previously been tagged by a
     * `ShimReferenceTagger`, restore the original `referencedFiles` array that does not have shim tags.
     */
    function untagAllTsFiles(program) {
        var e_1, _a;
        try {
            for (var _b = tslib_1.__values(program.getSourceFiles()), _c = _b.next(); !_c.done; _c = _b.next()) {
                var sf = _c.value;
                untagTsFile(sf);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
    }
    exports.untagAllTsFiles = untagAllTsFiles;
    /**
     * For those `ts.SourceFile`s in the `program` which have previously been tagged by a
     * `ShimReferenceTagger`, re-apply the effects of tagging by updating the `referencedFiles` array to
     * the tagged version produced previously.
     */
    function retagAllTsFiles(program) {
        var e_2, _a;
        try {
            for (var _b = tslib_1.__values(program.getSourceFiles()), _c = _b.next(); !_c.done; _c = _b.next()) {
                var sf = _c.value;
                retagTsFile(sf);
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_2) throw e_2.error; }
        }
    }
    exports.retagAllTsFiles = retagAllTsFiles;
    /**
     * Restore the original `referencedFiles` for the given `ts.SourceFile`.
     */
    function untagTsFile(sf) {
        if (sf.isDeclarationFile || !isExtended(sf)) {
            return;
        }
        var ext = sfExtensionData(sf);
        if (ext.originalReferencedFiles !== null) {
            sf.referencedFiles = ext.originalReferencedFiles;
        }
    }
    exports.untagTsFile = untagTsFile;
    /**
     * Apply the previously tagged `referencedFiles` to the given `ts.SourceFile`, if it was previously
     * tagged.
     */
    function retagTsFile(sf) {
        if (sf.isDeclarationFile || !isExtended(sf)) {
            return;
        }
        var ext = sfExtensionData(sf);
        if (ext.taggedReferenceFiles !== null) {
            sf.referencedFiles = ext.taggedReferenceFiles;
        }
    }
    exports.retagTsFile = retagTsFile;
});
